src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs

```diff
@@ -22,428 +22,698 @@ namespace NuGet.Commands
             {
-                return new MSBuildRestoreResult(targetsPath, propsPath, success: true);
+                propsXML = GenerateEmptyImportsFile();
             }
 
-            // Invalid msbuild projects should write out an msbuild error target
-            if (!targetGraphs.Any())
+            var files = new List<MSBuildOutputFile>()
             {
-                return new MSBuildRestoreResult(targetsPath, propsPath, success: false);
-            }
+                new MSBuildOutputFile(propsPath, propsXML),
+                new MSBuildOutputFile(targetsPath, targetsXML),
+            };
 
-            // Framework -> (targets, props)
-            var buildAssetsByFramework = new Dictionary<NuGetFramework, TargetsAndProps>();
+            return files;
+        }
 
-            // Get assets for each framework
-            foreach (var projectFramework in project.TargetFrameworks.Select(f => f.FrameworkName))
+        public static string ReplacePathsWithMacros(string path)
+        {
+            foreach (var macroName in MacroCandidates)
             {
-                var targetsAndProps =
-                    GetTargetsAndPropsForFramework(
-                        targetGraphs,
-                        repositories,
-                        context,
-                        request,
-                        includeFlagGraphs,
-                        projectFramework);
+                string macroValue = Environment.GetEnvironmentVariable(macroName);
+                if (!string.IsNullOrEmpty(macroValue)
+                    && path.StartsWith(macroValue, StringComparison.OrdinalIgnoreCase))
+                {
+                    path = $"$({macroName})" + $"{path.Substring(macroValue.Length)}";
+                }
+
+                break;
+            }
+
+            return path;
+        }
+
+        public static XDocument GenerateMultiTargetFrameworkWarning()
+        {
+            var doc = GenerateEmptyImportsFile();
+
+            doc.Root.Add(new XElement(Namespace + "Target",
+                        new XAttribute("Name", "EmitMSBuildWarning"),
+                        new XAttribute("BeforeTargets", "Build"),
+
+                        new XElement(Namespace + "Warning",
+                            new XAttribute("Text", Strings.MSBuildWarning_MultiTarget))));
+
+            return doc;
+        }
+
+        /// <summary>
+        /// Add standard properties to only props file if it exists, otherwise the targets.
+        /// </summary>
+        public static void AddNuGetPropertiesToFirstImport(IEnumerable<MSBuildOutputFile> files,
+            IEnumerable<string> packageFolders,
+            string repositoryRoot,
+            RestoreOutputType outputType,
+            bool success)
+        {
+            var firstImport = files.Where(file => file.Content != null)
+                .OrderByDescending(file => file.Path.EndsWith(PropsExtension) ? 1 : 0)
```

Comment:
For project.json there could be only a targets file, or no files that will be written out at all. For PackageReference this will always be props however.